### PR TITLE
[CELEBORN-450][HELM] Configurable volumes in the values.yaml

### PR DESCRIPTION
**Pull request by zwangsheng**
### What changes were proposed in this pull request?
Modify the chart of Celeborn.

Provide the mounts & volumes for users to configure in the `Values.yaml`.

And this change sets `emptyDirs` for default mount. This will allow the user to run the Celeborn cluster with minimal cost, while adding comments so that the users can make changes before run in production environment.

### Why are the changes needed?
Currently, the user cannot change the mount type unless change the template files, which is extremely complex.

### Does this PR introduce _any_ user-facing change?
Yes

### How was this patch tested?
Local Run `helm install celeborn .`

### CODE REVIEW

**Comment by zwangsheng** [comment]
> describe master pod
```yaml
Mounts:
      /mnt/rss_ratis/ from celeborn-master-volume (rw)
      /opt/celeborn/conf from celeborn-volume (ro)
      /var/run/secrets/kubernetes.io/serviceaccount from kube-api-access-vjl8r (ro)
Conditions:
  Type              Status
  Initialized       True 
  Ready             True 
  ContainersReady   True 
  PodScheduled      True 
Volumes:
  celeborn-volume:
    Type:      ConfigMap (a volume populated by a ConfigMap)
    Name:      celeborn-conf
    Optional:  false
  celeborn-master-volume:
    Type:       EmptyDir (a temporary directory that shares a pod's lifetime)
    Medium:     
    SizeLimit:  1Gi
```
> master pod log
```log
starting org.apache.celeborn.service.deploy.master.Master, logging to /opt/celeborn/logs/celeborn--org.apache.celeborn.service.deploy.master.Master-1-celeborn-master-0.out
OpenJDK 64-Bit Server VM warning: Cannot open file gc-master.out due to Permission denied

23/05/24 11:32:31,235 INFO [main] Dispatcher: Dispatcher numThreads: 4
23/05/24 11:32:31,316 INFO [main] TransportClientFactory: mode NIO threads 64
23/05/24 11:32:31,587 INFO [main] Utils: Successfully started service 'MasterSys' on port 9097.
23/05/24 11:32:31,708 INFO [main] Master: Metrics system enabled.
23/05/24 11:32:31,769 INFO [main] HttpServer: master: HttpServer started on port 9098.
23/05/24 11:32:31,779 INFO [main] Master: Master started.
23/05/24 11:32:32,407 INFO [dispatcher-event-loop-1] Master: Registered worker 
Host: 172.17.0.5
RpcPort: 37427
PushPort: 46233
FetchPort: 33547
ReplicatePort: 46067
SlotsUsed: 0
LastHeartbeat: 0
Disks: 
  DiskInfo0: DiskInfo(maxSlots: 1137, committed shuffles 0 shuffleAllocations: Map(), mountPoint: /mnt/disk1, usableSpace: 76348059648, avgFlushTime: 0, activeSlots: 0) status: HEALTHY dirs 
  DiskInfo1: DiskInfo(maxSlots: 1137, committed shuffles 0 shuffleAllocations: Map(), mountPoint: /mnt/disk0, usableSpace: 76348059648, avgFlushTime: 0, activeSlots: 0) status: HEALTHY dirs 
  DiskInfo2: DiskInfo(maxSlots: 1137, committed shuffles 0 shuffleAllocations: Map(), mountPoint: /mnt/disk2, usableSpace: 76348059648, avgFlushTime: 0, activeSlots: 0) status: HEALTHY dirs 
UserResourceConsumption: empty
WorkerRef: null
```

**Comment by pan3793** [comment]
please create a `values-it.yaml`, and keep the default `values.yaml` friendly to user.

File: charts/celeborn/values.yaml [hunk 1]
```diff
@@ -136,3 +136,38 @@ affinity:
...
+      emptyDir:
+        sizeLimit: 2Gi
+
+# When changing the mount point,
+# you need to change the corresponding configuration address like:
+# celeborn.ha.master.ratis.raft.server.storage.dir
+# celeborn.worker.storage.dirs
+volumeMounts:
+  master:
+    - name: celeborn-master-volume
```
Comment (RexXiong) [inline charts/celeborn/values.yaml:165]: The changes look good to me, and it would be better if CelebornConf(storage.dir/ratis dir) can modify accordingly

Comment (zwangsheng) [reply charts/celeborn/values.yaml:165]: Sure, i'm working on this

**Comment by zwangsheng** [comment]
Make a big changes, ready for review @RexXiong @pan3793, and the following is local test screenshoot for each type:
popo_2023-05-25  18-35-27

popo_2023-05-25  18-35-52

File: charts/celeborn/values.yaml [hunk 2]
```diff
@@ -36,10 +36,26 @@ workerReplicas: 5
...
+    dirs:
+      path: &worker-volume-dirs /mnt/disk1,/mnt/disk2,/mnt/disk3,/mnt/disk4
+      type: hostPath
+      sizeLimit: 1Gi
+
 # celeborn configurations
 celeborn:
-  celeborn.ha.master.ratis.raft.server.storage.dir: /mnt/rss_ratis/
-  celeborn.worker.storage.dirs: /mnt/disk1,/mnt/disk2,/mnt/disk3,/mnt/disk4
+  celeborn.ha.master.ratis.raft.server.storage.dir: *master-volume-ratis
```
Comment (pan3793) [inline charts/celeborn/values.yaml:57]: please use go template syntax, mixed-using yaml/go template var ref syntax for the same purpose increases the complication

File: charts/celeborn/templates/_helpers.tpl [hunk 3]
```diff
@@ -76,3 +76,48 @@ Create the name of the service account to use
 {{- default "default" .Values.serviceAccount.name }}
 {{- end }}
 {{- end }}
+
+{{- define "celeborn.master.volumeMounts" -}}
```
Comment (pan3793) [inline charts/celeborn/templates/_helpers.tpl:80]: if possible, move the logic to templates.

File: charts/celeborn/templates/configmap.yaml [hunk 4]
```diff
@@ -29,11 +29,12 @@ metadata:
 data:
   celeborn-defaults.conf: |-
     {{- $namespace := .Release.Namespace }}
-    celeborn.ha.enabled=true
     celeborn.master.endpoints={{ range until (.Values.masterReplicas |int) }}celeborn-master-{{ . }}.celeborn-master-svc.{{ $namespace }}.svc.{{ $.Values.cluster.name }}.local,{{ end }}
     {{- range until (.Values.masterReplicas |int) }}
     celeborn.ha.master.node.{{ . }}.host=celeborn-master-{{ . }}.celeborn-master-svc.{{ $namespace }}.svc.{{ $.Values.cluster.name }}.local
     {{- end }}
+    celeborn.ha.master.ratis.raft.server.storage.dir:{{ .Values.volumes.master.ratis.path }}
```
Comment (RexXiong) [inline charts/celeborn/templates/configmap.yaml:36]: the separator between key/value  need to be "=" rather than ":"？

**Comment by zwangsheng** [comment]
For Local Install Test

> celeborn worker pod spec
```yaml
Volumes:
  celeborn-volume:
    Type:      ConfigMap (a volume populated by a ConfigMap)
    Name:      celeborn-conf
    Optional:  false
  celeborn-worker-vol-0:
    Type:          HostPath (bare host directory volume)
    Path:          /mnt/disk1/worker
    HostPathType:  DirectoryOrCreate
  celeborn-worker-vol-1:
    Type:          HostPath (bare host directory volume)
    Path:          /mnt/disk2/worker
    HostPathType:  DirectoryOrCreate
  celeborn-worker-vol-2:
    Type:          HostPath (bare host directory volume)
    Path:          /mnt/disk3/worker
    HostPathType:  DirectoryOrCreate
  celeborn-worker-vol-3:
    Type:          HostPath (bare host directory volume)
    Path:          /mnt/disk4/worker
    HostPathType:  DirectoryOrCreate
```
> celeborn configmap
```properties
celeborn-defaults.conf:
----
celeborn.master.endpoints=celeborn-master-0.celeborn-master-svc.default.svc.cluster.local,celeborn-master-1.celeborn-master-svc.default.svc.cluster.local,celeborn-master-2.celeborn-master-svc.default.svc.cluster.local,
celeborn.ha.master.node.0.host=celeborn-master-0.celeborn-master-svc.default.svc.cluster.local
celeborn.ha.master.node.1.host=celeborn-master-1.celeborn-master-svc.default.svc.cluster.local
celeborn.ha.master.node.2.host=celeborn-master-2.celeborn-master-svc.default.svc.cluster.local
celeborn.ha.master.ratis.raft.server.storage.dir=/mnt/rss_ratis/master
celeborn.worker.storage.dirs=/mnt/disk1,/mnt/disk2,/mnt/disk3,/mnt/disk4
celeborn.application.heartbeat.timeout=120s
celeborn.ha.enabled=true
celeborn.master.metrics.prometheus.port=9098
celeborn.metrics.enabled=true
celeborn.push.stageEnd.timeout=120s
celeborn.rpc.dispatcher.numThreads=4
celeborn.rpc.io.clientThreads=64
celeborn.rpc.io.numConnectionsPerPeer=2
celeborn.rpc.io.serverThreads=64
celeborn.shuffle.chunk.size=8m
celeborn.worker.fetch.io.threads=32
celeborn.worker.flusher.buffer.size=256K
celeborn.worker.heartbeat.timeout=120s
celeborn.worker.metrics.prometheus.port=9096
celeborn.worker.monitor.disk.enabled=false
celeborn.worker.push.io.threads=32

```

File: charts/celeborn/templates/configmap.yaml (hunk 4 above)
Comment (zwangsheng) [reply charts/celeborn/templates/configmap.yaml]: Sure, fixed and also modify `volumes` from string to array.

File: charts/celeborn/values.yaml [hunk 5]
```diff
@@ -55,6 +77,7 @@ celeborn:
   celeborn.push.stageEnd.timeout: 120s
   celeborn.application.heartbeat.timeout: 120s
   celeborn.worker.heartbeat.timeout: 120s
+  celeborn.ha.enabled: true
```
Comment (pan3793) [inline charts/celeborn/values.yaml:80]: put ha switch ahead

File: charts/celeborn/templates/configmap.yaml [hunk 6]
```diff
@@ -29,11 +29,21 @@ metadata:
...
     celeborn.ha.master.node.{{ . }}.host=celeborn-master-{{ . }}.celeborn-master-svc.{{ $namespace }}.svc.{{ $.Values.cluster.name }}.local
     {{- end }}
+    {{- $dirs := .Values.volumes.master }}
+    celeborn.ha.master.ratis.raft.server.storage.dir={{ (index $dirs 0).path }}/master
+    {{- $path := "" }}
+    {{- range $worker := .Values.volumes.worker }}
+    {{- if eq $path "" }}
+    {{- $path = $worker.path }}
+    {{- else }}
+    {{- $path = ( list  $path $worker.path | join ",") }}
```
Comment (pan3793) [inline charts/celeborn/templates/configmap.yaml:43]: nit: an extra space

File: charts/celeborn/templates/worker-statefulset.yaml [hunk 7]
```diff
@@ -109,9 +108,15 @@ spec:
...
+          emptyDir:
+            sizeLimit: {{ $volume.size }}
+        {{- else if eq "hostPath" $volume.type }}
+        - name: celeborn-worker-vol-{{ $index }}
+          hostPath:
+            path: {{ $volume.path }}/worker
             type: DirectoryOrCreate
-          name: vol-{{ $index }}
-        {{- end }} 
+        {{- end }}
```
Comment (pan3793) [inline charts/celeborn/templates/worker-statefulset.yaml:121]: can we add an else branch to fail fast to guard other types?

**Comment by zwangsheng** [comment]
> Fast fail with another type

```log
Error: INSTALLATION FAILED: execution error at (celeborn/templates/master-statefulset.yaml:120:11): For now Celeborn Helm only support emptyDir or hostPath volume types
helm.go:84: [debug] execution error at (celeborn/templates/master-statefulset.yaml:120:11): For now Celeborn Helm only support emptyDir or hostPath volume types
INSTALLATION FAILED
main.newInstallCmd.func2

```

File: charts/celeborn/templates/worker-statefulset.yaml (hunk 7 above)
Comment (zwangsheng) [reply charts/celeborn/templates/worker-statefulset.yaml:123]: Sure, fixed.

File: charts/celeborn/templates/worker-statefulset.yaml [hunk 8]
```diff
@@ -109,9 +108,17 @@ spec:
...
+            sizeLimit: {{ $volume.size }}
+        {{- else if eq "hostPath" $volume.type }}
+        - name: celeborn-worker-vol-{{ $index }}
+          hostPath:
+            path: {{ $volume.path }}/worker
             type: DirectoryOrCreate
-          name: vol-{{ $index }}
-        {{- end }} 
+        {{- else }}
+        {{ fail "For now Celeborn Helm only support emptyDir or hostPath volume types" }}
```
Comment (pan3793) [inline charts/celeborn/templates/worker-statefulset.yaml:122]:
```suggestion
        {{ fail "Currently, Celeborn chart only supports 'emptyDir' and 'hostPath' volume types" }}
```

**Review by pan3793 — APPROVED** [review]

File: charts/celeborn/values.yaml [hunk 9]
```diff
@@ -36,10 +36,33 @@ workerReplicas: 5
...
   fsGroup: 10006
 
+# Current Celeborn support followings volume type:
+# - emptyDir
+# - hostPath
+# Note: size only works in emptyDir type
+# Celeborn Master will pick first volumes for store raft log
+volumes:
+  master:
+    - path: /mnt/rss_ratis
```
Comment (pan3793) [inline charts/celeborn/values.yaml:46]: A potential improvement(we can do it in another PR if the current implementation meets the requirements): separate `mountPath` and `hostPath`

Comment (zwangsheng) [reply charts/celeborn/values.yaml:46]: +1 separate `mountPath ` & `hostPath ` for more flexibility, i will do it in another PR.

**Review by RexXiong — APPROVED** [review]

**Comment by zwangsheng** [comment]
Thanks for review, after both approved, merged to  main.
